<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

_Author: Matt Brems_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="../images/schema.png" alt="schema" width="750"/>

In [1]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data.
engine = create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [2]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(*) 
FROM stores;
"""

pd.read_sql_query(sql, engine)

count
0   1973

**Prompt 1 Answer**: 1973

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [3]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT * 
FROM stores
ORDER BY name
LIMIT 4;
"""

pd.read_sql_query(sql, engine)

store                                name store_status  \
0   4986  218 Fuel Express & Chubby's Liquor            A   
1   3047          3047 Cub Foods / Iowa City            I   
2   3060               3060 Cub Foods / Ames            I   
3   3061         3061 Cub Foods / Sioux City            I   

                                       store_address address_info  
0  68 Monroe St\nFloyd, IA 50435\n(43.12249790500...         None  
1  855 Highway 1 West\nIowa City, IA 522460000\n(...         None  
2  3121 Grand Ave\nAmes, IA 500100000\n(42.053458...         None  
3  1732 Hamilton Blvd\nSioux City, IA 511030000\n...         None

**Prompt 2 Answer**: 3061 Cub Foods / Sioux City.

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price, bottle_qty
FROM sales
WHERE category_name = 'IMPORTED VODKA'
ORDER BY vendor ASC
LIMIT 3;
"""

pd.read_sql_query(sql, engine)

btl_price  bottle_qty
0     $8.99          12
1    $28.50           1
2    $28.50           6

**Prompt 3 Answer**: 
- Transaction 1: bottle price is \$8.99 and 12 bottles purchased.
- Transaction 2: bottle price is \$28.50 and 1 bottle purchased.
- Transaction 3: bottle price is \$28.50 and 6 bottles purchased.

### Prompt 4

What is the total amount of liquor sold in February 2015?

In [5]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT SUM(total)
FROM sales
WHERE date BETWEEN '2-1-2015' AND '2-28-2015';
"""
pd.read_sql_query(sql, engine)

sum
0  21295350.55

**Prompt 4 Answer**: The total amount of liquor sold in February 2015 is \$21,295,350.55.

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [6]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT vendor, COUNT(*) AS "Number of Transactions", SUM(bottle_qty) AS "Number of Bottles", SUM(total) AS "Total Revenue"
FROM sales
WHERE category_name = 'IMPORTED VODKA'
GROUP BY vendor
ORDER BY COUNT(*) DESC;
"""

pd.read_sql_query(sql, engine)

vendor  Number of Transactions  \
0       Pernod Ricard USA/Austin Nichols                   42338   
1                    Bacardi U.S.A. Inc.                   26441   
2                        Diageo Americas                   15725   
3        Constellation Wine Company Inc.                   12565   
4                        Jim Beam Brands                    5850   
5                            Stoli Group                    4443   
6                   Bacardi U.S.A., Inc.                    3344   
7                   Imperial Brands Inc.                    2799   
8                 Moet Hennessy USA Inc.                    2153   
9       Constellation Wine Company, Inc.                    1672   
10                               Proximo                    1628   
11                               MHW Ltd                    1159   
12                        Luxco-St Louis                    1128   
13                   Wilson Daniels Ltd.                     815   
14                 Imperial Brands, Inc.                     426   
15            Russian Standard Vodka USA                     415   
16                      Quadro Group LLC                     289   
17              Brown-Forman Corporation                     263   
18               Moet Hennessy USA, Inc.                     256   
19           WILLIAM GRANT AND SONS INC.                     232   
20                      Klin Spirits LLC                     218   
21                                 LUXCO                     140   
22           Russian Standard Vodka, USA                      71   
23                                  None                      62   
24          WILLIAM GRANT AND SONS, INC.                      56   
25                 A Hardy / U.S.A. Ltd.                      33   
26           W. J. Deutsch And Sons Ltd.                      24   
27                     Jinro America Inc                      21   
28              Firestarter Spirits Inc.                      20   
29      Western Spirits Beverage Co. LLC                      18   
30                             TY KU LLC                      17   
31          Park Street Imports LLC Voli                      12   
32  Shaw Ross International Importers LL                      10   
33                      Sazerac Co. Inc.                      10   
34                     Klin Spirits, LLC                       5   
35            The Patron Spirits Company                       4   
36                     Wine Intelligence                       3   
37                 Ti Beverage Group Ltd                       2   
38      Stanley Stawski Distributing Co.                       2   
39                     Phenix Brands LLC                       2   
40        Charles Jacquin Et. Cie., Inc.                       2   
41             Adamba Imports Int'l Inc.                       1   
42                A Hardy / U.S.A., Ltd.                       1   

    Number of Bottles  Total Revenue  
0              403450     7880365.88  
1              236047     5771242.07  
2              152038     3346425.42  
3              116053     1895005.51  
4              104119     1807097.23  
5               35513      832071.19  
6               30239      765347.31  
7               21033      265317.79  
8               10966      323126.54  
9               12956      200283.64  
10               9431      178212.03  
11               4493      109190.48  
12               8933      116914.19  
13               2396       84246.72  
14               3054       37950.20  
15               2421       40361.58  
16               1054       25928.40  
17               1730       23337.88  
18               1032       30783.53  
19               1119       20981.25  
20                784       15483.84  
21               2374       32425.32  
22                360        5400.00  
23                368        6009.44  
24                252        4725.00  
25                320      

**Prompt 5 Answer**: Pernod Ricard USA/Austin Nichols had 42,338 transactions, totalling 403,450 bottles and \$7,880,365.88 in revenue.

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [7]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price - state_btl_cost AS "Markup"
FROM sales;
"""

pd.read_sql_query(sql, engine)

Markup
0         $5.75
1         $5.75
2         $5.75
3         $8.58
4         $2.50
...         ...
3049908   $5.75
3049909  $10.00
3049910   $5.75
3049911   $5.75
3049912  $10.00

[3049913 rows x 1 columns]

**Prompt 6 Answer**: 

<img src="./prompt_6_screenshot.png" alt="prompt_6" height="800"/>

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [8]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT store, AVG((btl_price - state_btl_cost) / state_btl_cost) AS "Percentage Markup"
FROM sales
GROUP BY store
ORDER BY "Percentage Markup" DESC;
"""

pd.read_sql_query(sql, engine)

store  Percentage Markup
0      4013           0.525471
1      4024           0.521499
2      4266           0.518903
3      3833           0.515094
4      3660           0.514591
...     ...                ...
1347   9018           0.500000
1348   4784           0.499994
1349   4494           0.499909
1350   4534           0.499781
1351   4693           0.499214

[1352 rows x 2 columns]

**Prompt 7 Answer**: The store with the highest average markup percentage is store 4013, which has an average percentage markup of 52.5471%.

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [9]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT sales.store, AVG((sales.btl_price - sales.state_btl_cost) / sales.state_btl_cost) AS "Percentage Markup", stores.name, stores.store_address
FROM sales 
    LEFT JOIN stores
    ON sales.store = stores.store
GROUP BY sales.store, stores.name, stores.store_address
ORDER BY "Percentage Markup" DESC
LIMIT 5;
"""

pd.read_sql_query(sql, engine)

store  Percentage Markup                        name  \
0   4013           0.525471      Wal-Mart 0841 / Tipton   
1   4024           0.521499  Wal-Mart 1546 / Iowa Falls   
2   4266           0.518903  Wal-Mart 1683 / Shenandoah   
3   3833           0.515094    Wal-Mart 3394 / Atlantic   
4   3660           0.514591   Wal-Mart 2935 / Knoxville   

                                       store_address  
0  1126 Highway 38 North\nTipton, IA 527720000\n(...  
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...  
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...  
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....  
4  814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...

**Prompt 8 Answer**: The five stores with the highest markups are Wal-Mart 0841, Wal-Mart 1546, Wal-Mart 1683, Wal-Mart 3394, and Wal-Mart 2935.

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [10]:
# Use a SQL query to find the answer to the above prompt.

sql_subquery = """
SELECT * FROM
    (SELECT counties.county, (SUM(bottle_qty) / counties.population) AS "Bottles Sold per Capita", ROUND(AVG(products.bottle_size),1) AS "Average Bottle Size"
    FROM counties
        LEFT JOIN sales
            LEFT JOIN products
            ON sales.item = products.item_no
        ON counties.county = sales.county
    GROUP BY counties.county
    ORDER BY "Bottles Sold per Capita" DESC
    LIMIT 9) as df
ORDER BY "Average Bottle Size" DESC
LIMIT 1;
"""

sql = """
SELECT counties.county, (SUM(bottle_qty) / counties.population) AS "Bottles Sold per Capita", ROUND(AVG(products.bottle_size),1) AS "Average Bottle Size"
FROM counties
    LEFT JOIN sales
        LEFT JOIN products
        ON sales.item = products.item_no
    ON counties.county = sales.county
GROUP BY counties.county
ORDER BY "Bottles Sold per Capita" DESC
LIMIT 9
"""

pd.read_sql_query(sql, engine)

county  Bottles Sold per Capita  Average Bottle Size
0      Dickinson                       22                974.1
1     Black Hawk                       15                862.9
2           Polk                       15                892.6
3    Cerro Gordo                       14                986.4
4          Scott                       13                871.6
5        Johnson                       13                895.4
6           Linn                       12                886.3
7        Kossuth                       11               1041.0
8  Pottawattamie                       11                893.7

**Prompt 9 Answer**: Kossuth County sells the largest average bottle size (among the nine counties selling the most liquor bottles per capita), with an average bottle size of 1041.0 milliliters.

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [11]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(*)
FROM sales
WHERE CAST(total AS money) <> (btl_price * bottle_qty);
"""

pd.read_sql_query(sql, engine)

count
0      0

**Prompt 10 Answer**: There are no records where `total` is not equal to `btl_price * bottle_qty`.